File: Bank/Bank/src/main/java/com/example/Bank/service/CardValidationService.java
package com.example.Bank.service;

import com.example.Bank.model.Card;
import org.springframework.stereotype.Service;

import java.time.LocalDate;

@Service
public class CardValidationService {
    

    public boolean validateLuhn(String pan) {
        if (pan == null || pan.isEmpty()) {
            return false;
        }

        String digits = pan.replaceAll("\\D", "");
        Card tempCard = new Card();
        tempCard.setCardNumber(digits);
        return tempCard.isValid();
    }
    

    public boolean validateExpirationDate(String expirationDate) {
        if (expirationDate == null || !expirationDate.matches("\\d{2}/\\d{2}")) {
            return false;
        }
        
        try {
            String[] parts = expirationDate.split("/");
            int month = Integer.parseInt(parts[0]);
            int year = Integer.parseInt(parts[1]);

            if (month < 1 || month > 12) {
                return false;
            }

            LocalDate now = LocalDate.now();
            int currentYear = now.getYear() % 100;
            int currentMonth = now.getMonthValue();
            
            if (year < currentYear) {
                return false;
            }
            
            if (year == currentYear && month < currentMonth) {
                return false;
            }
            
            return true;
        } catch (Exception e) {
            return false;
        }
    }


    public String detectCardType(String pan) {
        if (pan == null || pan.isEmpty()) {
            return null;
        }

        String digits = pan.replaceAll("\\D", "");
        
        if (digits.startsWith("4")) {
            return "VISA";
        } else if (digits.startsWith("5") || digits.startsWith("2")) {
            return "MASTERCARD";
        }
        
        return null;
    }
}

File: Bank/bank-fr/src/app/services/card-validation.service.ts
import { Injectable } from '@angular/core';
import { CardType } from '../models/payment.models';

@Injectable({
  providedIn: 'root'
})
export class CardValidationService {
  
 
  validateLuhn(pan: string): boolean {
    if (!pan) {
      return false;
    }
    
    const digits = pan.replace(/\D/g, '');
    
    if (!digits || !/^\d+$/.test(digits)) {
      return false;
    }
    
    let sum = 0;
    let doubleDigit = false;
    
    for (let i = digits.length - 1; i >= 0; i--) {
      let digit = parseInt(digits[i]);
      
      if (doubleDigit) {
        digit *= 2;
        if (digit > 9) {
          digit -= 9;
        }
      }
      
      sum += digit;
      doubleDigit = !doubleDigit;
    }
    
    return sum % 10 === 0;
  }
  
  validateExpirationDate(expirationDate: string): boolean {
    if (!expirationDate || !/^\d{2}\/\d{2}$/.test(expirationDate)) {
      return false;
    }
    
    try {
      const [monthStr, yearStr] = expirationDate.split('/');
      const month = parseInt(monthStr);
      const year = parseInt(yearStr);
      
      if (month < 1 || month > 12) {
        return false;
      }
      
      const now = new Date();
      const currentYear = now.getFullYear() % 100;
      const currentMonth = now.getMonth() + 1;
      
      if (year < currentYear) {
        return false; 
      }
      
      if (year === currentYear && month < currentMonth) {
        return false; 
      }
      
      return true;
    } catch (e) {
      return false;
    }
  }
  

  detectCardType(pan: string): CardType {
    if (!pan) {
      return null;
    }
    
    const digits = pan.replace(/\D/g, '');
    
    if (digits.startsWith('4')) {
      return 'VISA';
    } else if (digits.startsWith('5') || digits.startsWith('2')) {
      return 'MASTERCARD';
    }
    
    return null;
  }
  

  formatCardNumber(pan: string): string {
    const digits = pan.replace(/\D/g, '');
    return digits.match(/.{1,4}/g)?.join(' ') || digits;
  }
  

  maskCardNumber(pan: string): string {
    const digits = pan.replace(/\D/g, '');
    if (digits.length <= 4) {
      return digits;
    }
    return '**** **** **** ' + digits.slice(-4);
  }
}






